File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
from collections import deque

for _ in range(int(input())):
    I = int(input())
    start_x, start_y = list(map(int, input().split()))
    end_x, end_y = list(map(int, input().split()))

    maps = [[0] * (I + 1) for _ in range(I + 1)]
    maps[start_x][start_y] = 1

    def bfs(start_x, start_y, end_x, end_y):
        queue = deque()
        queue.append((start_x, start_y))

        dx = [-1, -1, -2, -2, 1, 1, 2, 2]
        dy = [-2, 2, -1, 1, -2, 2, -1, 1]

        while queue:
            x, y = queue.popleft()

            if x == end_x and y == end_y:
                return maps[x][y] - 1

            for i in range(8):
                nx = x + dx[i]
                ny = y + dy[i]

                if 0 <= nx < I and 0 <= ny < I:
                    if maps[nx][ny] == 0:
                        queue.append((nx, ny))
                        maps[nx][ny] = maps[x][y] + 1


    print(bfs(start_x, start_y, end_x, end_y))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Queue;
import java.util.LinkedList;

public class Main {
    static int result;
    static int n;
    static int l;
    static int x1, x2, y1, y2;
    static int[][] dist;
    static boolean[][] visited;
    static int[] dx = {1, 2, 2, 1, -1, -2, -2, -1};
    static int[] dy = {2, 1, -1, -2, -2, -1, 1, 2};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            l = Integer.parseInt(br.readLine());
            dist = new int[l][l];
            visited = new boolean[l][l];

            StringTokenizer st = new StringTokenizer(br.readLine());
            x1 = Integer.parseInt(st.nextToken());
            y1 = Integer.parseInt(st.nextToken());

            st = new StringTokenizer(br.readLine());
            x2 = Integer.parseInt(st.nextToken());
            y2 = Integer.parseInt(st.nextToken());

            result = bfs(x1, y1, x2, y2);

            sb.append(result).append("\n");
        }

        System.out.println(sb);
    }

    public static int bfs(int startX, int startY, int endX, int endY) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{startX, startY});
        visited[startX][startY] = true;
        dist[startX][startY] = 0;

        while (!q.isEmpty()) {
            int[] now = q.poll();
            int x = now[0];
            int y = now[1];

            if (x == endX && y == endY) {
                return dist[x][y];
            }

            for (int i = 0; i < 8; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx >= 0 && ny >= 0 && nx < l && ny < l && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    dist[nx][ny] = dist[x][y] + 1;
                    q.add(new int[]{nx, ny});
                }
            }
        }

        return -1;
    }
}